### ISSUE TITLE
Jekyll v3 support - Liquid Exception: undefined method `cache_dir'

### ISSUE THREAD

**Issue by CraigChamberlain** [rbuchberger/jekyll_picture_tag#192]
Running into exceptions on first use of plugin.

Spent a long time trying to get imagemagik installed before giving up and attempting to use the docker image, jekyll/jekyll.

Has anyone experience of this error?

```
/srv/jekyll # bundle exec jekyll build --trace
Configuration file: /srv/jekyll/_config.yml
            Source: src
       Destination: public
 Incremental build: disabled. Enable with --incremental
      Generating... 
  Liquid Exception: undefined method `cache_dir' for #<Jekyll::Site:0x000055c8adb2aef0> in /_layouts/collection.html
bundler: failed to load command: jekyll (/usr/gem/bin/jekyll)
NoMethodError: undefined method `cache_dir' for #<Jekyll::Site:0x000055c8adb2aef0>
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/cache/base.rb:43:in `base_directory'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/cache/base.rb:55:in `filename'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/cache/base.rb:33:in `data'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/cache/base.rb:12:in `[]'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/source_image.rb:71:in `check_cache'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/source_image.rb:17:in `initialize'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/instructions/set.rb:63:in `new'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/instructions/set.rb:63:in `build_source_images'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/instructions/set.rb:34:in `source_images'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/router.rb:48:in `source_images'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/utils.rb:51:in `count_srcsets'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag/output_formats/auto.rb:7:in `new'
  /usr/gem/gems/jekyll_picture_tag-1.12.0/lib/jekyll_picture_tag.rb:68:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:91:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:160:in `block (2 levels) in render_segment'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:158:in `each'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:158:in `block in render_segment'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:123:in `stack'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:150:in `render_segment'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:79:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:82:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:208:in `block in render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:242:in `with_profiling'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:207:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:220:in `render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:30:in `block (2 levels) in render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:42:in `measure_bytes'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:29:in `block in render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:49:in `measure_time'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:28:in `render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/tags/include.rb:140:in `block in render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:123:in `stack'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/tags/include.rb:137:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:91:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:208:in `block in render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:242:in `with_profiling'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:207:in `render'
  /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:220:in `render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:30:in `block (2 levels) in render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:42:in `measure_bytes'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:29:in `block in render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:49:in `measure_time'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/liquid_renderer/file.rb:28:in `render!'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/renderer.rb:126:in `render_liquid'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/renderer.rb:195:in `render_layout'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/renderer.rb:158:in `place_in_layouts'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/renderer.rb:88:in `render_document'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/renderer.rb:62:in `run'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:479:in `render_regenerated'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:464:in `block (2 levels) in render_docs'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:463:in `each'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:463:in `block in render_docs'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:462:in `each_value'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:462:in `render_docs'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:191:in `render'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/site.rb:71:in `process'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/command.rb:28:in `process_site'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/commands/build.rb:65:in `build'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/commands/build.rb:36:in `process'
  /usr/gem/gems/jekyll-3.8.7/lib/jekyll/commands/build.rb:18:in `block (2 levels) in init_with_program'
  /usr/gem/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
  /usr/gem/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
  /usr/gem/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
  /usr/gem/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
  /usr/gem/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
  /usr/gem/gems/jekyll-3.8.7/exe/jekyll:15:in `<top (required)>'
  /usr/gem/bin/jekyll:23:in `load'
  /usr/gem/bin/jekyll:23:in `<top (required)>'
```

Gem File:
```
source "https://rubygems.org"

gem "jekyll", "~> 3.8"

group :jekyll_plugins do
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
    gem 'jekyll-seo-tag'
    gem "jekyll-sitemap", "~> 1.4"
    gem "jekyll_picture_tag", "~> 1.12"
end
```

**Comment by rbuchberger** [comment]
Oops! That might be my fault from the new update. I didn't test it against Jekyll v3, it's possible that the `cache_dir` setting I'm using is new in that version. As a workaround for now, you can either upgrade Jekyll to version 4+ or downgrade Jekyll Picture Tag to 1.10.2

I'm really sorry about that, I should have checked which version of Jekyll that was added in before using it.

**Comment by CraigChamberlain** [comment]
So I found that jekyll/jekyll does not have imagemagick installed.  I added it with the following commands

apk add imagemagick

Error persists so is firing before trying to trigger

**Comment by CraigChamberlain** [comment]
Hi No problem, at least you got back fast.  It's pushed be to learn a bit about docker too.  Update is taking a while, that could easily be the problem.  Will report back.

> Oops! That might be my fault from the new update. I didn't test it against Jekyll v3, it's possible that the `cache_dir` setting I'm using is new in that version. As a workaround for now, you can either upgrade Jekyll to version 4+ or downgrade Jekyll Picture Tag to 1.10.2
> 
> I'm really sorry about that, I should have checked which version of Jekyll that was added in before using it.

**Comment by CraigChamberlain** [comment]
Yes, it works fine with Jekyll updated not sure why I put the version lock in.  Would probably have been fine without that.

**Comment by rbuchberger** [comment]
Glad to hear it's working. I'll need to update the version spec at least, or figure out how to make it work with v3. Docker is fun... Though I feel  your pain, it's not exactly intuitive.

**Comment by rbuchberger** [comment]
Update - Here's my thinking on this issue:

JPT has been broken on Jekyll v3 since version 1.11 released back in July (because it uses v4's caching function, which I didn't realize was new.)  

This leaves me 2 choices: rewrite a caching feature for v3, or drop support. I can't officially drop support for Jekyll v3 without a new major release, which I've been trying to avoid. That said, there are a few other changes I'd like to make which would also require a new major version, so maybe it's time?

Basically, my time is limited and I'm not sure if it's best spent supporting old versions of Jekyll, considering that upgrading a site from v3 to v4 isn't too difficult anyway.

Tentatively, I think I'm going to go with the 'drop support in a new major version' option, unless I hear any strong objections.

**Comment by CraigChamberlain** [comment]
I vote you should bump the versions, feels more productive and I got sorted with your workaround upgrading Jekyll way easier than rewriting code.

You can make a pre-release on more or less the current state leave v.1.11 for longer term support.  I would have though that if someone is mad keen on keeping Jekyll on v.3 it's maybe on them to sort it out or make a donation if they have a business requirement holding them back from upgrading Jekyll.

**Comment by ChrisHammond** [comment]
I'm cool with dropping V3, clarity in your docs around this would help too! I have ran into this problem twice now and keep forgetting the resolution :D

**Comment by rbuchberger** [comment]
Thanks, glad to hear it. V2 is in the works now, hopefully we can get it finished in a few weeks. I'll look at adding better guidance in the meantime.
